File: apps/api/src/audit/audit.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { SupabaseService } from '../supabase/supabase.service';

export type AuditAction =
  | 'create'
  | 'update'
  | 'delete'
  | 'approve'
  | 'reject'
  | 'issue_id'
  | 'revoke_id'
  | 'view'
  | 'login';

export type AuditActorKind = 'officer' | 'citizen' | 'system';

export interface AuditEntry {
  actor_kind: AuditActorKind;
  actor_id?: string | null;
  action: AuditAction;
  entity_table: string;
  entity_id?: string | null;
  before_state?: unknown;
  after_state?: unknown;
  ip_address?: string | null;
  user_agent?: string | null;
}

@Injectable()
export class AuditService {
  private readonly logger = new Logger(AuditService.name);

  constructor(private readonly supabase: SupabaseService) {}

  async record(entry: AuditEntry): Promise<void> {
    const { error } = await this.supabase.admin.from('audit_log').insert({
      actor_kind: entry.actor_kind,
      actor_id: entry.actor_id ?? null,
      action: entry.action,
      entity_table: entry.entity_table,
      entity_id: entry.entity_id ?? null,
      before_state: entry.before_state ?? null,
      after_state: entry.after_state ?? null,
      ip_address: entry.ip_address ?? null,
      user_agent: entry.user_agent ?? null,
    });

    if (error) {
      // We never throw from audit — losing the request because of an audit
      // failure is worse than losing the audit row. Logged loudly so ops
      // can pick it up.
      this.logger.error(`Failed to write audit_log entry: ${error.message}`, error);
    }
  }
}

File: apps/api-dotnet/Audit/AuditService.cs
using System.Data;
using Microsoft.Data.SqlClient;
using Microsoft.EntityFrameworkCore;
using Sarh.Api.Data;

namespace Sarh.Api.Audit;

public sealed class AuditEntry
{
    public required string ActorKind { get; init; } // officer | citizen | system
    public Guid? ActorId { get; init; }
    public required string Action { get; init; }
    public required string EntityTable { get; init; }
    public Guid? EntityId { get; init; }
    public string? BeforeStateJson { get; init; }
    public string? AfterStateJson { get; init; }
    public string? IpAddress { get; init; }
    public string? UserAgent { get; init; }
}

// Append-only writer for the audit_log table. Inserts use a parameterised
// raw command rather than EF entity tracking — audit_log uses
// BIGINT IDENTITY for ordering and never participates in the request DbContext.
public sealed class AuditService(SarhDbContext db, ILogger<AuditService> log)
{
    public async Task RecordAsync(AuditEntry entry, CancellationToken ct)
    {
        try
        {
            var conn = (SqlConnection)db.Database.GetDbConnection();
            if (conn.State != ConnectionState.Open) await conn.OpenAsync(ct);
            await using var cmd = conn.CreateCommand();
            cmd.CommandText = @"
                INSERT INTO audit_log
                    (actor_kind, actor_id, action, entity_table, entity_id,
                     before_state, after_state, ip_address, user_agent)
                VALUES
                    (@actor_kind, @actor_id, @action, @entity_table, @entity_id,
                     @before_state, @after_state, @ip_address, @user_agent);";
            cmd.Parameters.Add(new SqlParameter("@actor_kind", SqlDbType.NVarChar, 16) { Value = entry.ActorKind });
            cmd.Parameters.Add(new SqlParameter("@actor_id", SqlDbType.UniqueIdentifier) { Value = (object?)entry.ActorId ?? DBNull.Value });
            cmd.Parameters.Add(new SqlParameter("@action", SqlDbType.NVarChar, 16) { Value = entry.Action });
            cmd.Parameters.Add(new SqlParameter("@entity_table", SqlDbType.NVarChar, 64) { Value = entry.EntityTable });
            cmd.Parameters.Add(new SqlParameter("@entity_id", SqlDbType.UniqueIdentifier) { Value = (object?)entry.EntityId ?? DBNull.Value });
            cmd.Parameters.Add(new SqlParameter("@before_state", SqlDbType.NVarChar, -1) { Value = (object?)entry.BeforeStateJson ?? DBNull.Value });
            cmd.Parameters.Add(new SqlParameter("@after_state", SqlDbType.NVarChar, -1) { Value = (object?)entry.AfterStateJson ?? DBNull.Value });
            cmd.Parameters.Add(new SqlParameter("@ip_address", SqlDbType.NVarChar, 45) { Value = (object?)entry.IpAddress ?? DBNull.Value });
            cmd.Parameters.Add(new SqlParameter("@user_agent", SqlDbType.NVarChar, -1) { Value = (object?)entry.UserAgent ?? DBNull.Value });
            await cmd.ExecuteNonQueryAsync(ct);
        }
        catch (Exception ex)
        {
            // Audit failures must never escalate — losing the request because
            // of an audit write is worse than losing the audit row.
            log.LogError(ex, "Failed to write audit_log entry for {Entity}/{Action}", entry.EntityTable, entry.Action);
        }
    }
}
